### PR TITLE
[EC-78] Block header validation

### PR DESCRIPTION
**Pull request by ntallar**
## Description
BlockHeader validation based on Section 4.4.4 of http://paper.gavwood.com/, as done by Geth in https://github.com/ethereum/go-ethereum/blob/master/core/block_validator.go#L206.

## Considerations
- Task EC-78 was added for doing some simple validation of the blocks broadcasted by other peers.
- So far only a simple Proof of Work validation is done that doesn't involve working with the DAG. This idea was obtained from EthereumJ (https://github.com/ethereum/ethereumj/blob/ad74b89f80b1edadba0c1e01c499166d0642c7d8/ethereumj-core/src/main/java/org/ethereum/validator/ProofOfWorkRule.java).

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/network/p2p/validators/BlockHeaderValidator.scala [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+  def validate(blockHeader: BlockHeader, parentHeader: BlockHeader): Either[BlockHeaderError, BlockHeader] = {
+    for {
+      _ <- validateExtraData(blockHeader)
+      _ <- validateTimestamp(blockHeader, parentHeader)
+      _ <- validateDifficulty(blockHeader, parentHeader)
+      _ <- validateGasUsed(blockHeader)
+      _ <- validateGasLimit(blockHeader, parentHeader)
+      _ <- validateNumber(blockHeader, parentHeader)
+      _ <- validatePoW(blockHeader)
+    } yield blockHeader
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/network/p2p/validators/BlockHeaderValidator.scala]: ❤️

File: src/test/scala/io/iohk/ethereum/network/p2p/validators/BlockHeaderValidatorSpec.scala [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+    ommersHash = ByteString(Hex.decode("1dcc4de8dec75d7aab85b567b6ccd41ad312451b948a7413f0a142fd40d49347")),
+    beneficiary = ByteString(Hex.decode("95f484419881c6e9b6de7fb3f8ad03763bd49a89")),
+    stateRoot = ByteString(Hex.decode("cddeeb071e2f69ad765406fb7c96c0cd42ddfc6ec54535822b564906f9e38e44")),
+    transactionsRoot = ByteString(Hex.decode("56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421")),
+    receiptsRoot = ByteString(Hex.decode("56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421")),
+    logsBloom = ByteString(Hex.decode("00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000")),
```
Comment (rtkaczyk) [inline src/test/scala/io/iohk/ethereum/network/p2p/validators/BlockHeaderValidatorSpec.scala]: How about: `ByteString(Hex.decode("00" * 256))` ?

File: src/main/scala/io/iohk/ethereum/domain/BlockHeader.scala [hunk 3]
```diff
@@ -40,3 +42,14 @@ case class BlockHeader(
...
+
+object BlockHeader {
+
+  def getEncodedWithoutNonce(blockHeader: BlockHeader): Array[Byte] = {
+    val rlpEncoded = headerRlpEncDec.encode(blockHeader) match {
+      case rlpList: RLPList => RLPList(rlpList.items.dropRight(2): _*)
+      case _ => throw new Exception("BlockHeader cannot be encoded without nonce and mixHash")
+    }
+    encode(rlpEncoded)
+  }
```
Comment (rtkaczyk) [inline src/main/scala/io/iohk/ethereum/domain/BlockHeader.scala]: Should this be unit tested?

File: src/test/scala/io/iohk/ethereum/network/p2p/validators/BlockHeaderValidatorSpec.scala (hunk 2 above)
Comment (ntallar) [inline src/test/scala/io/iohk/ethereum/network/p2p/validators/BlockHeaderValidatorSpec.scala]: I agree, will change that. I obtained that block header using @AlanVerbner 's tool and I missed that.

File: src/main/scala/io/iohk/ethereum/network/p2p/validators/BlockHeaderValidator.scala [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    else Left(HeaderNumberError)
+
+  /**
+    * Validates [[io.iohk.ethereum.domain.BlockHeader.nonce]] and [[io.iohk.ethereum.domain.BlockHeader.mixHash]] are correct
+    * based on validations stated in section 4.4.4 of http://paper.gavwood.com/
+    *
+    * @param blockHeader BlockHeader to validate.
+    * @return BlockHeader if valid, an [[HeaderPoWError]] otherwise
+    */
+  //FIXME: Simple PoW validation without using DAG
```
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/network/p2p/validators/BlockHeaderValidator.scala]: @mcsherrylabs Should we create a new task or we will address `FIXME` messages sometime within this phase?

Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/network/p2p/validators/BlockHeaderValidator.scala]: [[EC-88]](https://iohk.myjetbrains.com/youtrack/issue/EC-88) added.... feel free to comment on it..

File: src/main/scala/io/iohk/ethereum/crypto/package.scala [hunk 5]
```diff
@@ -26,6 +26,12 @@ package object crypto {
     digest.digest
   }
 
+  def sha512(input: Array[Byte]): Array[Byte] = {
```
Comment (mcsherrylabs) [inline src/main/scala/io/iohk/ethereum/crypto/package.scala]: @ntallar Have a look at [EC-89](https://iohk.myjetbrains.com/youtrack/issue/EC-89) shaxxx is the convention, but we can be more accurate. (that doesn't hold up merging this PR)

File: src/main/scala/io/iohk/ethereum/network/p2p/validators/BlockHeaderValidator.scala (hunk 4 above)
Comment (AlanVerbner) [inline src/main/scala/io/iohk/ethereum/network/p2p/validators/BlockHeaderValidator.scala]: @mcsherrylabs great! So this PR is ready to be merged then

**Pull request merged**